{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run X2K\n",
    "\n",
    "This notebook executes the eXpression2Kinases (X2K) pipeline, which takes up- and down- gene sets from each patient sample and infers the upstream TFs, intermediates and kinases likely regulating the observed differential expression. If phosphoproteomics are also available, we can find the overlap of kinases enriched in both transcriptomics and phosphoproteomics.\n",
    "\n",
    "CSVs containing the predicted TFs, intermediates, and kinases for each sample will be produced in the specified output folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import json\n",
    "import os\n",
    "import requests\n",
    "from scipy.stats import zscore\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parameters\n",
    "Customize analysis parameters below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_TFs = 20 # number of top-ranked TFs to add to the gene set\n",
    "augment = 'coexpression' # method of gene set expansion ['generif', 'tagger', 'enrichr', 'coexpression']\n",
    "n_genes = 100 # size of expanded gene set\n",
    "rank_method = 'Integrated--meanRank' # method of kinase ranking ['Integrated--meanRank', 'Integrated--topRank'] or the name of the database\n",
    "out_folder = \"results\" # folder to output results\n",
    "rank_threshold = 30 # cut-off of top-ranked kinases considered as enriched in a sample"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Functions to call the appropriate APIs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## load the expected baseline KEA rank distribution for each kinase--used for rank normalization\n",
    "kea_stats = pd.read_csv('data/KEA3_rank_stats.csv', index_col=0)\n",
    "## load the expected baseline ChEA rank distribution for each TF--used for rank normalization\n",
    "chea_stats = pd.read_csv('data/ChEA3_rank_stats.csv', index_col=0)\n",
    "\n",
    "def get_chea3_results(gene_set, query_name='', db=rank_method):\n",
    "    \"\"\" Inputs:\n",
    "    gene_set: list[]\n",
    "    query_name: str\n",
    "    db: ['Integrated--meanRank', 'Integrated--topRank'], ranking method for ChEA3 \n",
    "    Return: list[] of TFs by decreasing score (higher match) \"\"\" \n",
    "\n",
    "    ADDLIST_URL = 'https://maayanlab.cloud/chea3/api/enrich/'\n",
    "    payload = {\n",
    "        'gene_set': gene_set,\n",
    "        'query_name': query_name\n",
    "    }\n",
    "    \n",
    "    # Retrieve response\n",
    "    response = requests.post(ADDLIST_URL, data=json.dumps(payload))\n",
    "    if not response.ok:\n",
    "        raise Exception('Error analyzing gene list')\n",
    "    chea = json.loads(response.text)\n",
    "\n",
    "    # Save nodes\n",
    "    ## Lower scores indicate more relevancy to the transcription factor\n",
    "    chea_results = pd.DataFrame(chea[db])\n",
    "    chea_results = chea_results[['TF', 'Score']].set_index('TF')\n",
    "    return chea_results\n",
    "\n",
    "def get_kea3_results(gene_set, query_name='', db=rank_method):\n",
    "    \"\"\" Inputs:\n",
    "    gene_set: list[]\n",
    "    query_name: str\n",
    "    db: ['Integrated--meanRank', 'Integrated--topRank'], ranking method for KEA3 \n",
    "    Return: list[] of kinases by decreasing score (higher match) \"\"\" \n",
    "    \n",
    "    ADDLIST_URL = 'https://amp.pharm.mssm.edu/kea3/api/enrich/'\n",
    "    payload = {\n",
    "        'gene_set': gene_set,\n",
    "        'query_name': query_name\n",
    "    }\n",
    "    \n",
    "    # Retrieve response\n",
    "    response = requests.post(ADDLIST_URL, data=json.dumps(payload))\n",
    "    if not response.ok:\n",
    "        raise Exception('Error analyzing gene list')\n",
    "    data = json.loads(response.text)\n",
    "\n",
    "    df = pd.DataFrame([[k['TF'], k['Score']] for k in data[db]], columns=['Kinase','Score'])\n",
    "    df = df.set_index('Kinase')\n",
    "    return df\n",
    "\n",
    "def geneshot_set_augment(gene_list, similarity_type=augment, n_genes=n_genes):\n",
    "    \"\"\" Inputs:\n",
    "    gene_list: list[]\n",
    "    similarity_type: ['generif', 'tagger', 'coexpression', 'enrichr'], resource to do the augmentation\n",
    "    n_genes: int, number of the most similar genes to return\n",
    "    Return: list[] of augmented genes \"\"\"\n",
    "\n",
    "    GENESHOT_URL = 'https://maayanlab.cloud/geneshot/api/associate'\n",
    "    payload = {\n",
    "        \"gene_list\": gene_list,\n",
    "        \"similarity\": similarity_type \n",
    "        }\n",
    "    \n",
    "    # Retrieve response\n",
    "    response = requests.post(GENESHOT_URL, json=payload)\n",
    "    data = json.loads(response.text)\n",
    "  \n",
    "    # Return augmented genes\n",
    "    geneshot_dataframe = pd.DataFrame.from_dict(data['association'], orient=\"index\").drop(['topGenes', 'topScores'], axis=1).sort_values(by='simScore', ascending=False)\n",
    "    geneshot_dataframe = geneshot_dataframe.iloc[:n_genes]\n",
    "    return geneshot_dataframe.index.to_list()\n",
    "\n",
    "def get_g2n_results(geneset, subgraph_size=30, path_length=2):\n",
    "    # Inputs: \n",
    "    # 'geneset' : ['gene name'] - needs a list of strings\n",
    "    # 'databases' : ['database name'] - needs a list of strings,  \n",
    "    # options: iid, bioGRID, STRING, bioPlex 3.0, default - all of them \n",
    "    # 'subgraph_size' : int, - max size of the subgraph you want to produce\n",
    "    # 'path_length' : int, number of edges between target proteins \n",
    "\n",
    "    payload = {'geneset' : geneset, 'subgraph_size': str(subgraph_size), 'path_length': path_length} \n",
    "\n",
    "    # Retrieve response\n",
    "    res = requests.post(\"https://g2nkg.dev.maayanlab.cloud/api/knowledge_graph/ppi_kg\", json = payload)\n",
    "    results = res.json()\n",
    "\n",
    "    # Format results\n",
    "    nodes_data = {}\n",
    "    for i in results:\n",
    "        if i[\"data\"][\"kind\"] == 'Relation':\n",
    "            continue\n",
    "        else:\n",
    "            nodes_data[i[\"data\"][\"id\"]] = i[\"data\"]\n",
    "    \n",
    "    # Return nodes\n",
    "    nodes = pd.DataFrame.from_dict(nodes_data, orient=\"index\")\n",
    "    return nodes['label'].to_list()\n",
    "\n",
    "def chea_normalized(genes):\n",
    "    \"\"\" Return dataframe of rank-normalized TF scores \"\"\"\n",
    "    \n",
    "    results = get_chea3_results(gene_set=genes)\n",
    "    \n",
    "    # compute TF z-scores relative to the expected distribution\n",
    "    z_chea = pd.DataFrame({'zscore': (chea_stats['mean'].loc[results.index]-pd.to_numeric(results['Score'])).divide(chea_stats['std'].loc[results.index])}, \n",
    "                          index=results.index) # positive value indicates smaller score (higher rank) compared to average\n",
    "    return z_chea\n",
    "\n",
    "def kea_normalized(genes):\n",
    "    \"\"\" Return dataframe of rank-normalized kinase scores \"\"\"\n",
    "    \n",
    "    results = get_kea3_results(gene_set=genes)\n",
    "    \n",
    "    # compute kinase z-scores relative to the expected distribution\n",
    "    z_kea = pd.DataFrame({'zscore': (kea_stats['mean'].loc[results.index]-pd.to_numeric(results['Score'])).divide(kea_stats['std'].loc[results.index])}, \n",
    "                          index=results.index) # positive value indicates smaller score (higher rank) compared to average\n",
    "    return z_kea"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify the gene set files to use as input:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "up_filename = \"data/CPTAC/RNA_all.z_V1.1_isoform_adjusted_top500.tsv\" # replace with your own up gene set file\n",
    "dn_filename = \"data/CPTAC/RNA_all.z_V1.1_isoform_adjusted_bot500.tsv\" # replace with your own down gene set file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The input files should be formatted such that each column (sample ID) contains the gene set for that sample. If the TF files are already prepared, steps 1 & 2 should be skipped and the normalized TF scores can be uploaded here: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chea_up = pd.read_csv('data/CPTAC/ChEA3_10_cancer_type_meanRank_df_top500_baseline_normalized.tsv', sep='\\t', index_col=0)\n",
    "chea_dn = pd.read_csv('data/CPTAC/ChEA3_10_cancer_type_meanRank_df_bot500_baseline_normalized.tsv', sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Up- and down- gene sets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load gene sets\n",
    "\n",
    "In this example, we use a gene set size of 500 as the analysis input for each sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>C3L-00004</th>\n",
       "      <th>C3L-00010</th>\n",
       "      <th>C3L-00011</th>\n",
       "      <th>C3L-00026</th>\n",
       "      <th>C3L-00079</th>\n",
       "      <th>C3L-00088</th>\n",
       "      <th>C3L-00096</th>\n",
       "      <th>C3L-00097</th>\n",
       "      <th>C3L-00103</th>\n",
       "      <th>C3L-00183</th>\n",
       "      <th>...</th>\n",
       "      <th>17OV036</th>\n",
       "      <th>17OV039</th>\n",
       "      <th>17OV040</th>\n",
       "      <th>18OV001</th>\n",
       "      <th>20OV005</th>\n",
       "      <th>26OV002</th>\n",
       "      <th>26OV008</th>\n",
       "      <th>26OV009</th>\n",
       "      <th>26OV011</th>\n",
       "      <th>26OV013</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OR12D1</td>\n",
       "      <td>OR52L1</td>\n",
       "      <td>ACTL8</td>\n",
       "      <td>RP11-345F18.1</td>\n",
       "      <td>INA</td>\n",
       "      <td>POU1F1</td>\n",
       "      <td>NLRP4</td>\n",
       "      <td>H3-4</td>\n",
       "      <td>OR52E8</td>\n",
       "      <td>SRMS</td>\n",
       "      <td>...</td>\n",
       "      <td>SPATA19</td>\n",
       "      <td>GAGE12J</td>\n",
       "      <td>IGHD5-12</td>\n",
       "      <td>CMTM5</td>\n",
       "      <td>HHATL</td>\n",
       "      <td>DEFA5</td>\n",
       "      <td>UCN3</td>\n",
       "      <td>DMRTC2</td>\n",
       "      <td>TRGJP</td>\n",
       "      <td>SLC26A8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CXCR4</td>\n",
       "      <td>OR56A1</td>\n",
       "      <td>AICDA</td>\n",
       "      <td>VGF</td>\n",
       "      <td>EFCAB3</td>\n",
       "      <td>PRDM14</td>\n",
       "      <td>CCDC105</td>\n",
       "      <td>L34079.2</td>\n",
       "      <td>OR52E6</td>\n",
       "      <td>RP11-520P18.5</td>\n",
       "      <td>...</td>\n",
       "      <td>CYP3A7</td>\n",
       "      <td>H2AB3</td>\n",
       "      <td>CD5L</td>\n",
       "      <td>GLYATL1</td>\n",
       "      <td>ALLC</td>\n",
       "      <td>DEFA6</td>\n",
       "      <td>RBL1</td>\n",
       "      <td>TM4SF20</td>\n",
       "      <td>RP11-231I13.2</td>\n",
       "      <td>CLEC1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AC016577.1</td>\n",
       "      <td>OR9Q2</td>\n",
       "      <td>MRAP</td>\n",
       "      <td>TTC6</td>\n",
       "      <td>RTL1</td>\n",
       "      <td>KRT72</td>\n",
       "      <td>DPPA2</td>\n",
       "      <td>KRTAP19-1</td>\n",
       "      <td>ECT2L</td>\n",
       "      <td>HBG1</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV12-3</td>\n",
       "      <td>MAGEA6</td>\n",
       "      <td>SPO11</td>\n",
       "      <td>MYH6</td>\n",
       "      <td>C20orf202</td>\n",
       "      <td>RSPO2</td>\n",
       "      <td>PRB2</td>\n",
       "      <td>DCDC2C</td>\n",
       "      <td>WFDC9</td>\n",
       "      <td>CRNN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FPR3</td>\n",
       "      <td>OR56A5</td>\n",
       "      <td>DEFB135</td>\n",
       "      <td>HS3ST6</td>\n",
       "      <td>SP9</td>\n",
       "      <td>KRT73</td>\n",
       "      <td>GTSF1</td>\n",
       "      <td>NPIPA2</td>\n",
       "      <td>OR52N1</td>\n",
       "      <td>CCDC27</td>\n",
       "      <td>...</td>\n",
       "      <td>TRAV5</td>\n",
       "      <td>AHSG</td>\n",
       "      <td>FAM124B</td>\n",
       "      <td>CTXND1</td>\n",
       "      <td>CYP4F8</td>\n",
       "      <td>BSX</td>\n",
       "      <td>SERPINB12</td>\n",
       "      <td>SOHLH1</td>\n",
       "      <td>MRGPRD</td>\n",
       "      <td>CLEC2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CSF2RA</td>\n",
       "      <td>MYF6</td>\n",
       "      <td>CALB2</td>\n",
       "      <td>XKR3</td>\n",
       "      <td>PRPH</td>\n",
       "      <td>KIR3DL1</td>\n",
       "      <td>SLC22A31</td>\n",
       "      <td>RPS10</td>\n",
       "      <td>GP9</td>\n",
       "      <td>TAFA3</td>\n",
       "      <td>...</td>\n",
       "      <td>CAVIN4</td>\n",
       "      <td>MAGEA3</td>\n",
       "      <td>HSD3B1</td>\n",
       "      <td>C4orf45</td>\n",
       "      <td>ACSL6</td>\n",
       "      <td>ZSCAN5B</td>\n",
       "      <td>KLHL1</td>\n",
       "      <td>TAF7L</td>\n",
       "      <td>ANGPTL5</td>\n",
       "      <td>VSTM2B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1083 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    C3L-00004 C3L-00010 C3L-00011      C3L-00026 C3L-00079 C3L-00088  \\\n",
       "1      OR12D1    OR52L1     ACTL8  RP11-345F18.1       INA    POU1F1   \n",
       "2       CXCR4    OR56A1     AICDA            VGF    EFCAB3    PRDM14   \n",
       "3  AC016577.1     OR9Q2      MRAP           TTC6      RTL1     KRT72   \n",
       "4        FPR3    OR56A5   DEFB135         HS3ST6       SP9     KRT73   \n",
       "5      CSF2RA      MYF6     CALB2           XKR3      PRPH   KIR3DL1   \n",
       "\n",
       "  C3L-00096  C3L-00097 C3L-00103      C3L-00183  ...   17OV036  17OV039  \\\n",
       "1     NLRP4       H3-4    OR52E8           SRMS  ...   SPATA19  GAGE12J   \n",
       "2   CCDC105   L34079.2    OR52E6  RP11-520P18.5  ...    CYP3A7    H2AB3   \n",
       "3     DPPA2  KRTAP19-1     ECT2L           HBG1  ...  TRBV12-3   MAGEA6   \n",
       "4     GTSF1     NPIPA2    OR52N1         CCDC27  ...     TRAV5     AHSG   \n",
       "5  SLC22A31      RPS10       GP9          TAFA3  ...    CAVIN4   MAGEA3   \n",
       "\n",
       "    17OV040  18OV001    20OV005  26OV002    26OV008  26OV009        26OV011  \\\n",
       "1  IGHD5-12    CMTM5      HHATL    DEFA5       UCN3   DMRTC2          TRGJP   \n",
       "2      CD5L  GLYATL1       ALLC    DEFA6       RBL1  TM4SF20  RP11-231I13.2   \n",
       "3     SPO11     MYH6  C20orf202    RSPO2       PRB2   DCDC2C          WFDC9   \n",
       "4   FAM124B   CTXND1     CYP4F8      BSX  SERPINB12   SOHLH1         MRGPRD   \n",
       "5    HSD3B1  C4orf45      ACSL6  ZSCAN5B      KLHL1    TAF7L        ANGPTL5   \n",
       "\n",
       "   26OV013  \n",
       "1  SLC26A8  \n",
       "2   CLEC1B  \n",
       "3     CRNN  \n",
       "4   CLEC2A  \n",
       "5   VSTM2B  \n",
       "\n",
       "[5 rows x 1083 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "up_genes = pd.read_csv(up_filename, sep='\\t')\n",
    "dn_genes = pd.read_csv(dn_filename, sep='\\t')\n",
    "ids = up_genes.columns.to_list()\n",
    "\n",
    "up_genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Submit up- and down- genes for TFEA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, the TFs predicted to regulate the up- and down- gene sets for each sample are ranked using ChEA3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute normalized TF enrichment scores\n",
    "chea_up = {}\n",
    "chea_dn = {}\n",
    "\n",
    "with tqdm(total=len(ids)) as pbar:\n",
    "    pbar.set_description(\"Computing normalized ChEA scores...\")\n",
    "    for id in ids:\n",
    "        up_results = chea_normalized(up_genes[id].to_list())\n",
    "        dn_results = chea_normalized(dn_genes[id].to_list())\n",
    "        if len(chea_up) == 0:\n",
    "            TFs = up_results.zscore.index.to_list()\n",
    "        chea_up[id] = up_results.zscore.loc[TFs]\n",
    "        chea_dn[id] = dn_results.zscore.loc[TFs]\n",
    "        pbar.update(1)\n",
    "\n",
    "chea_up = pd.DataFrame(chea_up, index=TFs)\n",
    "chea_dn = pd.DataFrame(chea_dn, index=TFs)\n",
    "chea_up.to_csv(f\"{out_folder}/up_chea_zscores.csv\")\n",
    "chea_dn.to_csv(f\"{out_folder}/dn_chea_zscores.csv\")\n",
    "\n",
    "chea_up.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Gene set augmentation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get intermediates from up- and down- TFs\n",
    "up_TFs = {}\n",
    "up_intermediates = {}\n",
    "with tqdm(total=len(ids)) as pbar:\n",
    "    pbar.set_description(\"Augmenting TFs from up-genes...\")\n",
    "    for id in ids:\n",
    "        up_TFs[id] = chea_up[id].sort_values(ascending=False).index[:n_TFs].to_list()\n",
    "        up_intermediates[id] = geneshot_set_augment(gene_list=up_TFs[id])\n",
    "        pbar.update(1)\n",
    "up_TFs = pd.DataFrame(up_TFs)\n",
    "up_intermediates = pd.DataFrame(up_intermediates)\n",
    "up_TFs.to_csv(f\"{out_folder}/up_TFs.csv\", index=False)\n",
    "up_intermediates.to_csv(f\"{out_folder}/up_intermediates.csv\", index=False)\n",
    "\n",
    "dn_TFs = {}\n",
    "dn_intermediates = {}\n",
    "with tqdm(total=len(ids)) as pbar:\n",
    "    pbar.set_description(\"Augmenting TFs from down-genes...\")\n",
    "    for id in ids:\n",
    "        dn_TFs[id] = chea_dn[id].sort_values(ascending=False).index[:n_TFs].to_list()\n",
    "        dn_intermediates[id] = geneshot_set_augment(gene_list=dn_TFs[id])\n",
    "        pbar.update(1)\n",
    "dn_TFs = pd.DataFrame(dn_TFs)\n",
    "dn_intermediates = pd.DataFrame(dn_intermediates)\n",
    "dn_TFs.to_csv(f\"{out_folder}/dn_TFs.csv\", index=False)\n",
    "dn_intermediates.to_csv(f\"{out_folder}/dn_intermediates.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter TFs out of intermediates\n",
    "TFs = chea_up.index.to_list()\n",
    "for file in os.listdir(out_folder):\n",
    "    data = {}\n",
    "    if file.endswith(\"_intermediates.csv\"):\n",
    "        f = pd.read_csv(f\"{out_folder}/{file}\")\n",
    "        for id in f.columns:\n",
    "            newlist = []\n",
    "            for x in f[id].to_list():\n",
    "                if x not in TFs: newlist.append(x)\n",
    "            data[id] = newlist\n",
    "        pd.DataFrame(data).to_csv(f\"{out_folder}/filtered_{file}\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Kinase enrichment analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute normalized kinase enrichment scores\n",
    "up_kinases = {}\n",
    "dn_kinases = {}\n",
    "x2k_up = {}\n",
    "x2k_dn = {}\n",
    "\n",
    "with tqdm(total=len(ids)) as pbar:\n",
    "    pbar.set_description(\"Computing normalized KEA scores...\")\n",
    "    for id in ids:\n",
    "        up_geneset = list(set(list(up_TFs[id]) + list(up_intermediates[id])))\n",
    "        dn_geneset = list(set(list(dn_TFs[id]) + list(dn_intermediates[id])))\n",
    "        up_results = kea_normalized(up_geneset)\n",
    "        dn_results = kea_normalized(dn_geneset)\n",
    "        up_kinases[id] = up_results.sort_values(by=\"zscore\", ascending=False).index.to_list()\n",
    "        dn_kinases[id] = dn_results.sort_values(by=\"zscore\", ascending=False).index.to_list()\n",
    "        if len(x2k_up) == 0:\n",
    "            KINASES = up_results.zscore.index.to_list()\n",
    "        x2k_up[id] = up_results.zscore.loc[KINASES]\n",
    "        x2k_dn[id] = dn_results.zscore.loc[KINASES]\n",
    "        pbar.update(1)\n",
    "\n",
    "pd.DataFrame(up_kinases).to_csv(f\"{out_folder}/up_kinases.csv\", index=False)\n",
    "pd.DataFrame(dn_kinases).to_csv(f\"{out_folder}/dn_kinases.csv\", index=False)\n",
    "x2k_up = pd.DataFrame(x2k_up, index=KINASES)\n",
    "x2k_dn = pd.DataFrame(x2k_dn, index=KINASES)\n",
    "x2k_up.to_csv(f\"{out_folder}/up_x2k_zscores.csv\")\n",
    "x2k_dn.to_csv(f\"{out_folder}/dn_x2k_zscores.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check kinase recovery"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If phosphoproteomic data is available, we can check the agreement between the X2K kinase predictions and the direct phosphoproteomic kinase predictions. Uploade the normalized kinase scores for each sample here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kea_up = pd.read_csv('data/CPTAC/KEA3_V2_10_cancer_type_meanRank_df_top500_baseline_normalized.tsv', sep='\\t', index_col=0)\n",
    "kea_dn = pd.read_csv('data/CPTAC/KEA3_V2_10_cancer_type_meanRank_df_bot500_baseline_normalized.tsv', sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids = x2k_up.columns.intersection(kea_up.columns).to_list()\n",
    "\n",
    "# Kinases enriched in both X2K and phospho- pipelines\n",
    "rec_kinases = {}\n",
    "for id in ids:\n",
    "    up = x2k_up[id].sort_values(ascending=False).index.to_list()\n",
    "    up_phospho = kea_up[id].sort_values(ascending=False).index.to_list()\n",
    "    rec_kinases[id] = list(set(up[:rank_threshold]) & set(up_phospho[:rank_threshold]))\n",
    "pd.DataFrame(rec_kinases).to_csv(f\"{out_folder}/recovered_up_phospho_up_genes.csv\", index=False)\n",
    "\n",
    "rec_kinases = {}\n",
    "for id in ids:\n",
    "    dn = x2k_dn[id].sort_values(ascending=False).index.to_list()\n",
    "    dn_phospho = kea_dn[id].sort_values(ascending=False).index.to_list()\n",
    "    rec_kinases[id] = list(set(dn[:rank_threshold]) & set(dn_phospho[:rank_threshold]))\n",
    "pd.DataFrame(rec_kinases).to_csv(f\"{out_folder}/recovered_dn_phospho_dn_genes.csv\", index=False)\n",
    "\n",
    "rec_kinases = {}\n",
    "for id in ids:\n",
    "    dn = x2k_dn[id].sort_values(ascending=False).index.to_list()\n",
    "    up_phospho = kea_up[id].sort_values(ascending=False).index.to_list()\n",
    "    rec_kinases[id] = list(set(dn[:rank_threshold]) & set(up_phospho[:rank_threshold]))\n",
    "pd.DataFrame(rec_kinases).to_csv(f\"{out_folder}/recovered_up_phospho_dn_genes.csv\", index=False)\n",
    "\n",
    "rec_kinases = {}\n",
    "for id in ids:\n",
    "    up = x2k_up[id].sort_values(ascending=False).index.to_list()\n",
    "    dn_phospho = kea_dn[id].sort_values(ascending=False).index.to_list()\n",
    "    rec_kinases[id] = list(set(up[:rank_threshold]) & set(dn_phospho[:rank_threshold]))\n",
    "pd.DataFrame(rec_kinases).to_csv(f\"{out_folder}/recovered_dn_phospho_up_genes.csv\", index=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
